### PR TITLE
Fix legend yerr

### LINKED ISSUE #40044: BUG: plotting with `secondary_y` and then plotting another line results in wrong color for (right) line
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd

df = pd.DataFrame({'a': [1, 1], 'b': [2, 2]})
df2 = pd.DataFrame({'d': [2.5, 2.5]})

ax = df.plot(legend=True, secondary_y='b')
df2.plot(legend=False, ax=ax);
```

#### Problem description

Here's what we get:

image

#### Expected Output

The colours in the legend should be correct, i.e.:

image

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 272435073c70b4421f72184663837f4864601f11
python           : 3.8.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-66-generic
Version          : #74-Ubuntu SMP Wed Jan 27 22:54:38 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.3.0.dev0+841.g272435073c
numpy            : 1.19.5
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 49.6.0.post20201009
Cython           : 0.29.21
pytest           : 6.2.2
hypothesis       : 6.2.0
sphinx           : 3.5.1
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.6.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.19.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.8.5
fastparquet      : 0.5.0
gcsfs            : 0.7.2
matplotlib       : 3.3.3
numexpr          : 2.7.2
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : 2.0.0
pyxlsb           : None
s3fs             : 0.5.2
scipy            : 1.6.0
sqlalchemy       : 1.3.23
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.52.0

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] closes #40044
- [x] xref #39522
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

I spent several hours on this, and I don't think there's a simple solution at the moment (though I'd love to be proved wrong!)

Background: 

in Nov 2017, #18222 was opened, in which it was noted that marker info isn't preserved in legends. This is because we were getting legend handles using `leg.legendHandles`:

https://github.com/pandas-dev/pandas/blob/b623a9df3cf3d69fbdddab0f02a03ad69281c8b0/pandas/plotting/_matplotlib/core.py#L574

Due to [an open issue in matplotlib](https://github.com/matplotlib/matplotlib/issues/11357), `leg.legendHandles` doesn't preserve marker info. So, in #27808, the code was updated to use `ax.get_legend_handles_labels`:

https://github.com/pandas-dev/pandas/blob/5d3b492743e199e82e3018bb943545181ed0018f/pandas/plotting/_matplotlib/core.py#L604

This preserves marker info _but_ has the problem that it may contain additional objects which aren't visible in the legend. The result of this is the permutations observed in #39522 . 

I think that having wrong colours in the legend, as in #39522 and #40044, is _far_ worse than the legend not preserving marker info. Solving everything together doesn't currently seem doable, at least not until https://github.com/matplotlib/matplotlib/issues/11357 is solved in matplotlib.

So my suggestion is to revert some of the changes in #27808, so that at least the colors are right, and when the matplotlib issue is addressed, then the marker info can be preserved.

----

#39522 on master:

image

on this branch:

image

---

#40044 on master:

image

on this branch:

image

### CODE REVIEW

**Comment by jreback** [comment]
is  #39522 closed? (or partial)

File: pandas/tests/plotting/frame/test_frame_legend.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import pytest
+
```
Comment (jreback) [inline pandas/tests/plotting/frame/test_frame_legend.py:2]: can you move other legend tests here? (if we have them)

**Comment by MarcoGorelli** [comment]
@jreback it's only a partial fix of #39522 , as the marker info isn't preserved

> can you move other legend tests here? (if we have them)

Sure, though for ease of review, should I do that in a follow-up PR?

**Comment by jreback** [comment]
> Sure, though for ease of review, should I do that in a follow-up PR?

either way

**Comment by MarcoGorelli** [comment]
OK, have moved those tests over in this pr @jreback

**Comment by jreback** [comment]
thanks @MarcoGorelli very nice!

**Pull request merged**